File: web/projects/portal/src/app/composer/gui/composer-client.service.ts
/*
 * inetsoft-web - StyleBI is a business intelligence web application.
 * Copyright © 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affrero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import { Injectable, NgZone } from "@angular/core";
import { Observable ,  Subject } from "rxjs";
import { StompClientService } from "../../common/viewsheet-client";
import { StompClientConnection } from "../../../../../shared/stomp/stomp-client-connection";

@Injectable()
export class ComposerClientService {
   private connection: StompClientConnection;
   private connecting: boolean = false;
   private editComposerAssetSubject: Subject<any> = new Subject<any>();

   /**
    * Creates a new instance of <tt>ComposerClientService</tt>.
    */
   constructor(private client: StompClientService, private zone: NgZone) {
   }

   /**
    * Connects to the server.
    */
   public connect(): void {
      if(!this.connecting && !this.connection) {
         this.client.connect("../vs-events").subscribe(
            (connection) => {
               this.connecting = false;
               this.connection = connection;
               this.subscribe();
            },
            (error: any) => {
               this.connecting = false;
               console.error("Failed to connect to server: ", error);
               this.editComposerAssetSubject.error(error);
            }
         );
      }
   }

   /**
    * Subscribes the to the composer-client topic.
    */
   private subscribe(): void {
      this.connection.subscribe("/user/composer-client", (message) => {
         this.zone.run(() => {
            let event: any = JSON.parse(message.frame.body);
            this.editComposerAssetSubject.next(event);
         });
      });
   }

   private unsubscribe(): void {
      this.connection.send("/user/composer-client/leave", null, null);
   }

   /**
    * Disconnects from the server.
    */
   public disconnect(): void {
      if(this.editComposerAssetSubject) {
         this.editComposerAssetSubject.complete();
         this.editComposerAssetSubject = null;
      }

      if(this.connection) {
         this.unsubscribe();
         this.connection.disconnect();
         this.connection = null;
      }
   }

   /**
    * The observable from which the edit composer asset events are received.
    */
   public get editAsset(): Observable<any> {
      return this.editComposerAssetSubject.asObservable();
   }
}

File: core/src/main/java/inetsoft/web/viewsheet/service/ComposerClientService.java
/*
 * This file is part of StyleBI.
 * Copyright (C) 2024  InetSoft Technology
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package inetsoft.web.viewsheet.service;

import inetsoft.sree.internal.cluster.Cluster;
import inetsoft.sree.internal.cluster.DistributedMap;
import jakarta.annotation.PostConstruct;
import org.springframework.messaging.simp.stomp.StompHeaderAccessor;
import org.springframework.stereotype.Service;

import java.util.Map;

@Service
public class ComposerClientService {
   @PostConstruct
   public void init() {
      composerClients = Cluster.getInstance().getReplicatedMap(getClass().getName() + ".sessions");
   }

   public void removeFromSessionList(StompHeaderAccessor headers) {
      String simpSessionId = headers.getSessionId();

      if(simpSessionId != null) {
         composerClients.remove(simpSessionId);
      }
   }

   public String getFirstSimpSessionId(String httpSessionId) {
      return composerClients.entrySet().stream()
         .filter(e -> httpSessionId.equals(e.getValue()))
         .map(Map.Entry::getKey)
         .findFirst()
         .orElse(null);
   }

   public void setSessionID(String httpSessionId, String simpSessionId) {
      composerClients.put(simpSessionId, httpSessionId);
   }

   public static final String COMMANDS_TOPIC = "/composer-client";
   // key = simpSessionId, value = httpSessionId (cluster-wide replicated map)
   private DistributedMap<String, String> composerClients;
}
